src/connectedk8s/azext_connectedk8s/custom.py

```diff
@@ -1015,3 +1056,961 @@ def upgrade_agents(cmd, client, resource_group_name, cluster_name, kube_config=N
+                                summary='Exception while loading kubernetes configuration')
+        raise CLIError('Exception while loading kubernetes configuration.' + str(ex))
+
+    if context_name is not None:
+        addition['contexts'][0]['name'] = context_name
+        addition['contexts'][0]['context']['cluster'] = context_name
+        addition['clusters'][0]['name'] = context_name
+        addition['current-context'] = context_name
+
+    # rename the admin context so it doesn't overwrite the user context
+    for ctx in addition.get('contexts', []):
+        try:
+            if ctx['context']['user'].startswith('clusterAdmin'):
+                admin_name = ctx['name'] + '-admin'
+                addition['current-context'] = ctx['name'] = admin_name
+                break
+        except (KeyError, TypeError):
+            continue
+
+    if addition is None:
+        telemetry.set_exception(exception='Failed to load additional configuration', fault_type=consts.Failed_To_Load_K8s_Configuration_Fault_Type,
+                                summary='failed to load additional configuration from {}'.format(addition_file))
+        raise CLIError('failed to load additional configuration from {}'.format(addition_file))
+
+    if existing is None:
+        existing = addition
+    else:
+        handle_merge(existing, addition, 'clusters', replace)
+        handle_merge(existing, addition, 'users', replace)
+        handle_merge(existing, addition, 'contexts', replace)
+        existing['current-context'] = addition['current-context']
+
+    # check that ~/.kube/config is only read- and writable by its owner
+    if platform.system() != 'Windows':
+        existing_file_perms = "{:o}".format(stat.S_IMODE(os.lstat(existing_file).st_mode))
+        if not existing_file_perms.endswith('600'):
+            logger.warning('%s has permissions "%s".\nIt should be readable and writable only by its owner.',
+                           existing_file, existing_file_perms)
+
+    with open(existing_file, 'w+') as stream:
+        try:
+            yaml.safe_dump(existing, stream, default_flow_style=False)
+        except Exception as e:
+            telemetry.set_exception(exception=e, fault_type=consts.Failed_To_Merge_Kubeconfig_File,
+                                    summary='Exception while merging the kubeconfig file')
+            raise CLIError('Exception while merging the kubeconfig file.' + str(e))
+
+    current_context = addition.get('current-context', 'UNKNOWN')
+    msg = 'Merged "{}" as current context in {}'.format(current_context, existing_file)
+    print(msg)
+
+
+def handle_merge(existing, addition, key, replace):
+    if not addition[key]:
+        return
+    if existing[key] is None:
+        existing[key] = addition[key]
+        return
+
+    for i in addition[key]:
+        for j in existing[key]:
+            if i['name'] == j['name']:
+                if replace or i == j:
+                    existing[key].remove(j)
```

Comment:
Removing an element during the loop would cause unexpected behavior. In Python 3.8 it would move the iterator to the next one and then the iterator continues to move to the next one in the loop, so the element after the one that's removed will be skipped. You can run the below code to verify:
```
list=[1,2,3,4]
for n in list:
     list.remove(n)
print(list) # [2,4]
```

Reply:
Thank you so much for pointing this out. I think we never observed this because mostly one deletion happens to the list. But in case of multiple deletions, this issue would definitely arise. I have fixed the code and now we are not using list.remove().

Reply:
Providing some explanation here on why I removed iteration over addition list: handle_merge() function is called by merge_kubernetes_configurations() function in the code. The latter assumes that addition list has only one object. This can be confirmed from following lines in that function:
if context_name is not None:
        addition['contexts'][0]['name'] = context_name
        addition['contexts'][0]['context']['cluster'] = context_name
        addition['clusters'][0]['name'] = context_name
        addition['current-context'] = context_name
Hence, removing the outer loop to just use the first object made sense to me.